### ISSUE TITLE
Org-Roam capability

### ISSUE THREAD

**Issue by Firehawke** [jsonreeder/vscode-org-mode#161]
Might be worth looking into eventually integrating the capability of [Org-roam](https://org-roam.readthedocs.io/en/latest/) as that's almost certainly going to be on a lot of peoples' minds.

**Comment by ssimontis** [comment]
Hello @Firehawke, Org-roam is personally one of my favorite Emacs extensions. I am going to commit to this task unless anyone has any objections. I will post some design documentation here within the next 24 hours I hope.

**Comment by ssimontis** [comment]
For a basic version, I forsee the following tasks being necessary:

- [ ] Create roam template
- [ ] Add configuration settings
- [ ] Allow creation of links in between pages
- [ ] Allow definition of metadata on individual pages
- [ ] Search interface for notes
- [ ] GraphViz generation and display of note graph

**Comment by suvidh** [comment]
COOL I would really appreciate org-roam!

**Comment by ssimontis** [comment]
@PGilbertSchmitt @ajtoo @jsonreeder 

Is this something you would like integrated with the current extension, or would you prefer that I make another extension and we work together to promote the VS Code ORG family or perhaps make my extension a submodule?

I'm potentially interested in bringing over more ORG extensions, so I want to ensure that such an approach is manageable and doesn't result in a chaotic codebase.

**Comment by suvidh** [comment]
I am not the best of programmer.
If you integrate Org-roam with the current extension then there would be very few -albeit weird- people who will not like roam capabilities. They will not have a choice. 
If you make  a complete different module then maybe it would result into chaotic codebase. Since, you could have to repeat code. Moreover then Org-roam would probably still be dependent on Org.
I think keeping Org as main extension, Org-Roam and other org extension as submodule seems like the best option.

**Comment by sipi** [comment]
Hi @ssimontis, 
PGilbertSchmitt, ajtoo, jsonreeder are no longer very active on this project, and I try to ensure continuity. So, I know little about org-roam but as I understand it, main features are hyperlinks and search capabilities in order to built Wiki like note system.

From my point of view, hyperlinks and search capabilities are great features.
Having note templates to quickly create the skeleton of a new notes is a great feature two.
Regarding "note graph", it's interesting although not essential in my opinion.

So, if I take the liberty of rewriting the tasks (by priority in my point of view):

- [ ] Allow creation of links in between pages
  - [ ] Add configuration settings?
- [ ] Search interface for notes
  - [ ] Allow definition of metadata on individual pages
  - [ ] Add configuration settings?
- [ ] Create roam template
  - [ ] Add configuration settings?
- [ ] GraphViz generation and display of note graph
  - [ ] Add configuration settings

Does my understanding of inter-task dependencies seem consistent to you?

So "full integration" or "submodule" ? I don't have a definitive answer.
In my point of view, theses features are good. So I'll probably be ok with a full integration but perhaps some people doesn't need theses features… So submodule is more likely to be accepted by all. But I don't know how "submodule extension" works, I have to learn about to fully understand what it implies.

**Comment by ssimontis** [comment]
Hey @sipi,

Thanks for taking over this project! I agree with your prioritization. The graph feature was part of what I felt differentiated org-roam from some of the other similar extensions, but it can definitely be the last thing on my list.

I spent some time thinking things over regarding the submodule, and as nice as it seems, I don't think it would be feasible after going through the logic in my head. It would be a disaster trying to integrate with the existing codebase because I couldn't easily re-use any code, and if we start adding more and more extensions, we will end up with a very confusing repo where locating the exact location of a piece of code becomes an exercise in frustration. I will do my best to keep things modularized, and I will study the settings available in the original org-roam LISP code to see what functionality can be enabled/disabled. If that doesn't proove sufficient, I have some other ideas for how we can configure this and future extensions.

TL;dr: I will keep my code clean so we can make an executive decision later with minimum disruption if need be.

**Comment by theherk** [comment]
@ssimontis did anything come of this. I would love to see this feature. org-roam and [org-roam-ui](https://github.com/org-roam/org-roam-ui) would be killer.
